File: app/src/main/java/com/shapun/screentranslater/service/ScreenTranslateService.java
package com.shapun.screentranslater.service;

import android.accessibilityservice.AccessibilityService;
import android.accessibilityservice.AccessibilityServiceInfo;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.res.Configuration;
import android.graphics.PixelFormat;
import android.graphics.Rect;
import android.os.Build;
import android.util.Log;
import android.view.GestureDetector;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;
import android.view.accessibility.AccessibilityEvent;
import android.view.accessibility.AccessibilityNodeInfo;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;
import androidx.core.content.ContextCompat;
import androidx.core.graphics.ColorUtils;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.google.android.material.color.MaterialColors;
import com.shapun.screentranslater.R;
import com.shapun.screentranslater.activity.TranslateTextActivity;
import com.shapun.screentranslater.adapter.SelectedTextsAdapter;
import com.shapun.screentranslater.preferences.Preferences;
import com.shapun.screentranslater.util.AccessibilityServiceUtils;
import com.shapun.screentranslater.util.Utils;
import com.shapun.screentranslater.widget.MarkableView;

import java.util.ArrayList;
import java.util.List;

public class ScreenTranslateService extends AccessibilityService
        implements View.OnTouchListener, GestureDetector.OnGestureListener {
    private static final String TAG = "ScreenTranslateService";
    public static final int NOTIFICATION_ID = 100101;
    public static final String ACTION_HIDE_NOTIFICATION = "hide_notification";
    public static final String ACTION_SHOW_NOTIFICATION = "show_notification";
	public static final String ACTION_START_TRANSLATE_MODE ="start_translate_mode";
    // Variables must be set to null after use
    private WindowManager mWindowManager;
    private MarkableView mMarkableView;
    private WindowManager.LayoutParams mContainerParams;
    private GestureDetector mGestureDetector;
    private CoordinatorLayout mCoordinatorLayout;
    private TextView mTextViewSelectedCount;
    private RecyclerView mRecyclerViewSelectedText;
    private ArrayList<String> mSelectedTexts;
    private ArrayList<AccessibilityNodeInfo> mSelectedNodes;
    private SelectedTextsAdapter mSelectedTextAdapter;
    private int mSelectedWindowId;

    @Override
    protected void onServiceConnected() {
        Log.i(TAG, "Service Connected");
        setTheme(R.style.AppTheme);
        setServiceInfo(new AccessibilityServiceInfo());
        showNotification();
    }

    @Override
    public boolean onSingleTapUp(MotionEvent ev) {
        mSelectedWindowId = getRootInActiveWindow().getWindowId();
        List<AccessibilityNodeInfo> list = new ArrayList<>();
        getTexts(list, getRootInActiveWindow(), (int) ev.getRawX(), (int) ev.getRawY());
        if (list.size() == 0) return false;
        AccessibilityNodeInfo node = list.get(list.size() - 1);
        if (mSelectedNodes.contains(node)) {
            removeNode(node);
        } else {
            addNode(node);
        }
        return true;
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        if (mSelectedWindowId != 0) {
            stopTranslateService();
            startTranslateService();
        }
    }

    private void showNotification() {
        final String CHANNEL = "main";
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            CharSequence name = "Translate Service ";
            String description = "This is primary notification used to display translate options";
            int importance = NotificationManager.IMPORTANCE_MIN;
            NotificationChannel channel = new NotificationChannel(CHANNEL, name, importance);
            channel.setDescription(description);

            NotificationManager notificationManager = getSystemService(NotificationManager.class);
            notificationManager.createNotificationChannel(channel);
        }
        Intent intent = new Intent(this, ScreenTranslateService.class);
        intent.setAction(ACTION_START_TRANSLATE_MODE);
		int color = MaterialColors.getColor(
                        this,
                        android.R.attr.colorPrimary,
                        ContextCompat.getColor(this,R.color.colorPrimary));
        NotificationCompat.Builder builder =
                new NotificationCompat.Builder(this, CHANNEL)
                        .setContentTitle(getString(R.string.app_name))
                        .setContentText("Tap to translate or copy text on screen")
                        .setContentIntent(PendingIntent.getService(this, 0, intent, 0))
                        .setSmallIcon(R.drawable.default_image)
                        .setOngoing(true)
                        .setColor(color)
                        .setPriority(NotificationCompat.PRIORITY_LOW)
                        .setAutoCancel(false);
        NotificationManagerCompat notificationManager = NotificationManagerCompat.from(this);
        notificationManager.notify(NOTIFICATION_ID, builder.build());
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent == null) return super.onStartCommand(intent, flags, startId);
				switch (intent.getAction()) {
                    case ACTION_START_TRANSLATE_MODE:
                        startTranslateService();
                        break;
                    case ACTION_HIDE_NOTIFICATION:
                        ((NotificationManager) getSystemService(NOTIFICATION_SERVICE))
                                .cancel(NOTIFICATION_ID);
                        break;
                    case ACTION_SHOW_NOTIFICATION:
						setTheme(Preferences.getTheme(this));
                        showNotification();
                        break;
                    default:
                        break;
                }
        return START_STICKY;
    }

    private void startTranslateService() {
        mSelectedTexts = new ArrayList<>();
        mSelectedNodes = new ArrayList<>();
		setTheme(Preferences.getTheme(this));
        mWindowManager = (WindowManager) getSystemService(WINDOW_SERVICE);
	    setUpContainer();
        setUpBottomSheet();
        setUpRecyclerView();
        Intent it = new Intent(Intent.ACTION_CLOSE_SYSTEM_DIALOGS);
        sendBroadcast(it);
		try{
			mWindowManager.addView(mMarkableView, mContainerParams);
        }catch(Exception e){
			boolean permissionGiven = AccessibilityServiceUtils.isAccessibilityServiceEnabled(
                                this, ScreenTranslateService.class);
			Utils.toast(this,String.valueOf(getWindows()));
			Log.e(TAG,e.toString());
			Log.i(TAG,String.valueOf(permissionGiven));
			
		}
		mMarkableView.setAlpha(0);
        mMarkableView.animate().alpha(1).setDuration(1000);
        AccessibilityServiceInfo info = new AccessibilityServiceInfo();
        info.eventTypes = AccessibilityEvent.WINDOWS_CHANGE_ACTIVE;
        info.feedbackType = AccessibilityServiceInfo.FEEDBACK_ALL_MASK;
        setServiceInfo(info);
    }

    private void setUpBottomSheet() {
        View view =
                LayoutInflater.from(this).inflate(R.layout.dialog_selected_texts, mMarkableView);
        mTextViewSelectedCount = view.findViewById(R.id.selected_text_count);
        View sheetContent = view.findViewById(R.id.sheet_content);
        BottomSheetBehavior.from(sheetContent).setState(BottomSheetBehavior.STATE_COLLAPSED);
        ImageView img_close = view.findViewById(R.id.img_close);

        img_close.setOnClickListener(v -> stopTranslateService());
        view.findViewById(R.id.img_copy)
                .setOnClickListener(
                        v -> {
                            if (mSelectedTexts.isEmpty()) {
                                Utils.showSnackbar(
                                        mMarkableView,
                                        "No texts were selected . Click on texts you want copy");
                                return;
                            }
                            StringBuilder sb = new StringBuilder();
                            for (String str : mSelectedTexts) {
                                sb.append(str);
                                sb.append("\n");
                            }
                            Utils.copyToClipboard(this, sb);
                            Utils.showSnackbar(mMarkableView, "Copied to clipboard");
                        });
        view.findViewById(R.id.img_translate)
                .setOnClickListener(
                        v -> {
                            if (mSelectedTexts.isEmpty()) {
                                Utils.showSnackbar(
                                        mMarkableView,
                                        "No texts were selected . Click on texts you want translate");
                                return;
                            }
                            Intent intent = TranslateTextActivity.newIntent(this, mSelectedTexts);
                            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                            startActivity(intent);
                            stopTranslateService();
                        });

        view.findViewById(R.id.img_select_all)
                .setOnClickListener(
                        v -> {
                            List<AccessibilityNodeInfo> list = new ArrayList<>();
                            AccessibilityServiceUtils.getAllNodesWithText(
                                    list, this, getRootInActiveWindow());
                            for (AccessibilityNodeInfo info : list) {
                                addNode(info);
                            }
                        });
    }

    private void setUpRecyclerView() {
        mRecyclerViewSelectedText = mMarkableView.findViewById(R.id.recyclerview);
        mRecyclerViewSelectedText.setLayoutManager(new LinearLayoutManager(this));
        mSelectedTextAdapter = new SelectedTextsAdapter(mSelectedTexts);
        mRecyclerViewSelectedText.setAdapter(mSelectedTextAdapter);
        mSelectedTextAdapter.setOnCloseClickListener(
                pos -> {
                    removeNode(mSelectedNodes.get(pos));
                });
    }

    private void setUpContainer() {
        mMarkableView = new MarkableView(this);
        mContainerParams = new WindowManager.LayoutParams();
        mContainerParams.type = WindowManager.LayoutParams.TYPE_ACCESSIBILITY_OVERLAY;
        mContainerParams.format = PixelFormat.TRANSLUCENT;
		mContainerParams.flags |= WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE
                        | WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS
                        | WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS;
        mContainerParams.width = WindowManager.LayoutParams.MATCH_PARENT;
        mContainerParams.height = WindowManager.LayoutParams.MATCH_PARENT;
        int border = Utils.dpToPx(this, 5);
        int color = MaterialColors.getColor(
                        this,
                        android.R.attr.colorControlNormal,
                ContextCompat.getColor(this,R.color.colorControlNormal));
        mMarkableView.setHighilightColor(ColorUtils.setAlphaComponent(color, 255 / 4));
        mMarkableView.setBackgroundColor(0);
        mMarkableView.setBorderWidth(border);
        mMarkableView.setBorderColor(color);
        mMarkableView.setPadding(border, border, border, border);
        mMarkableView.setOnTouchListener(this);
        mGestureDetector = new GestureDetector(this, this);
    }

    private void addNode(AccessibilityNodeInfo accessibilityNodeInfo) {
        if (mSelectedNodes.contains(accessibilityNodeInfo)) return;
        mSelectedNodes.add(accessibilityNodeInfo);
        Rect rect = new Rect();
        accessibilityNodeInfo.getBoundsInScreen(rect);
        mMarkableView.addRect(rect);
        mSelectedTexts.add(accessibilityNodeInfo.getText().toString());
        mSelectedTextAdapter.notifyItemInserted(mSelectedTexts.size() - 1);
        refreshSelectedCount();
    }

    private void removeNode(AccessibilityNodeInfo node) {
        int index = mSelectedNodes.indexOf(node);
        if (index != -1) {
            Rect rect = new Rect();
            node.getBoundsInScreen(rect);
            mMarkableView.removeRect(rect);
            mSelectedNodes.remove(index);
            mSelectedTexts.remove(index);
            mSelectedTextAdapter.notifyItemRemoved(index);
            mSelectedTextAdapter.notifyItemRangeChanged(index, mSelectedTextAdapter.getItemCount());
            refreshSelectedCount();
        }
    }

    private void refreshSelectedCount() {
        TextView textView = mTextViewSelectedCount;
        String prefix =
                mSelectedTexts.size() == 0 ? "No text" : String.valueOf(mSelectedTexts.size());
        mTextViewSelectedCount.setText(prefix + "  selected");
    }

    @Nullable
    private void getTexts(
            List<AccessibilityNodeInfo> list, AccessibilityNodeInfo node, int x, int y) {
        if (node == null) return;
        if (node.getText() != null) {
            Rect rect = new Rect();
            node.getBoundsInScreen(rect);
            if (rect.contains(x, y)) {
                list.add(node);
            }
        }
        for (int i = 0; i < node.getChildCount(); i++) {
            AccessibilityNodeInfo child = node.getChild(i);
            getTexts(list, child, x, y);
        }
    }

    public void stopTranslateService() {
        mWindowManager.removeView(mMarkableView);
        setServiceInfo(new AccessibilityServiceInfo());
        mSelectedWindowId = 0;
        mWindowManager = null;
        mMarkableView = null;
        mContainerParams = null;
        mGestureDetector = null;
        mCoordinatorLayout = null;
        mTextViewSelectedCount = null;
        mRecyclerViewSelectedText = null;
        mSelectedTexts = null;
        mSelectedNodes = null;
    }

    @Override
    public void onAccessibilityEvent(AccessibilityEvent event) {
        if (event != null) {
            if (mSelectedWindowId != 0) {
                // Window changed sostop the service
                if (mSelectedWindowId != event.getWindowId()) {
                    stopTranslateService();
                }
            }
        }
    }

    @Override
    public void onDestroy() {
		super.onDestroy();
        NotificationManager mNotificationManager =
                (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
        mNotificationManager.cancel(NOTIFICATION_ID);
        Log.v(TAG, "Service Destroyed");
		Utils.toast(this,"destroyed");
    }

    @Override
    public void onInterrupt() {
        Log.v(TAG, "onInterrupt");
    }

    @Override
    public boolean onTouch(View v, MotionEvent e) {
        return mGestureDetector.onTouchEvent(e);
    }

    @Override
    public void onLongPress(MotionEvent e) {}

    @Override
    public boolean onScroll(MotionEvent e1, MotionEvent e2, float distanceX, float distanceY) {
        return false;
    }

    @Override
    public boolean onFling(MotionEvent e1, MotionEvent e2, float velocityX, float velocityY) {
        return false;
    }

    @Override
    public void onShowPress(MotionEvent e) {}

    @Override
    public boolean onDown(MotionEvent e) {
        return false;
    }
}

File: app/src/main/java/com/shapun/screentranslater/service/ScreenTranslateService.kt
package com.shapun.screentranslater.service

import android.accessibilityservice.AccessibilityService
import android.accessibilityservice.AccessibilityServiceInfo
import android.annotation.SuppressLint
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Intent
import android.content.res.Configuration
import android.graphics.PixelFormat
import android.graphics.Rect
import android.os.Build
import android.util.Log
import android.view.*
import android.view.View.OnTouchListener
import android.view.accessibility.AccessibilityEvent
import android.view.accessibility.AccessibilityNodeInfo
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.graphics.ColorUtils
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.shapun.screentranslater.R
import com.shapun.screentranslater.activity.TranslateTextActivity.Companion.newIntent
import com.shapun.screentranslater.adapter.SelectedTextsAdapter
import com.shapun.screentranslater.databinding.DialogSelectedTextsBinding
import com.shapun.screentranslater.preferences.Preferences
import com.shapun.screentranslater.util.AccessibilityServiceUtils
import com.shapun.screentranslater.util.UiUtils
import com.shapun.screentranslater.util.Utils

class ScreenTranslateService : AccessibilityService(), OnTouchListener,
    GestureDetector.OnGestureListener {
    //These Variables must be set to null after use
    private var mWindowManager: WindowManager? = null
    //private var mContainerParams: WindowManager.LayoutParams? = null
    private var mGestureDetector: GestureDetector? = null
    private var binding: DialogSelectedTextsBinding? = null
    private var mSelectedTexts: ArrayList<String>? = null
    private var mSelectedNodes: ArrayList<AccessibilityNodeInfo>? = null
    private var mSelectedTextAdapter: SelectedTextsAdapter? = null
    private var mSelectedWindowId = 0

    override fun onServiceConnected() {
        Log.i(TAG, "Service Connected")
        setTheme(R.style.AppTheme)
        serviceInfo = AccessibilityServiceInfo()
        showNotification()
    }

    override fun onSingleTapUp(ev: MotionEvent): Boolean {
        mSelectedWindowId = rootInActiveWindow.windowId
        val list = getTexts(rootInActiveWindow, ev.rawX.toInt(), ev.rawY.toInt())
        if (list.size == 0) return false
        val node = list[list.size - 1]
        if (mSelectedNodes!!.contains(node)) {
            removeNode(node)
        } else {
            addNode(node)
        }
        return true
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        if (mSelectedWindowId != 0) {
            stopTranslateService()
            startTranslateService()
        }
    }

    @SuppressLint("LaunchActivityFromNotification")
    private fun showNotification() {
        val channelName = "main"
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val name = "Translate Service "
            val description = "This is primary notification used to display translate options"
            val importance = NotificationManager.IMPORTANCE_MIN
            val channel = NotificationChannel(channelName, name, importance)
            channel.description = description
            val notificationManager = getSystemService(NotificationManager::class.java)
            notificationManager.createNotificationChannel(channel)
        }
        val intent = Intent(this, ScreenTranslateService::class.java)
        intent.action = ACTION_START_TRANSLATE_MODE
        val color = UiUtils.getColorPrimary(this)
        val serviceIntent = PendingIntent.getService(
            this,
            0,
            intent,
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT
            } else {
                PendingIntent.FLAG_UPDATE_CURRENT
            }
        )
        val builder = NotificationCompat.Builder(this, channelName)
            .setContentTitle(getString(R.string.app_name))
            .setContentText("Tap to translate or copy text on screen")
            .setContentIntent(serviceIntent)
            .setSmallIcon(R.drawable.default_image)
            .setOngoing(true)
            .setColor(color)
            .setPriority(NotificationCompat.PRIORITY_LOW)
            .setAutoCancel(false)
        NotificationManagerCompat.from(this).notify(NOTIFICATION_ID, builder.build())
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        when (intent.action) {
            ACTION_START_TRANSLATE_MODE -> startTranslateService()
            ACTION_HIDE_NOTIFICATION ->
                (getSystemService(NOTIFICATION_SERVICE) as NotificationManager).cancel(NOTIFICATION_ID)
            ACTION_SHOW_NOTIFICATION -> {
                setTheme(Preferences.getTheme(this))
                showNotification()
            }
            else -> {}
        }
        return START_STICKY
    }

    @SuppressLint("MissingPermission")
    private fun startTranslateService() {
        mSelectedTexts = ArrayList()
        mSelectedNodes = ArrayList()
        setTheme(Preferences.getTheme(this))
        mWindowManager = getSystemService(WINDOW_SERVICE) as WindowManager
        setUpContainer()
        setUpBottomSheet()
        setUpRecyclerView()
        binding!!.markableView.alpha = 0f
        binding!!.markableView.animate().alpha(1f).duration = 1000
        val info = AccessibilityServiceInfo()
        info.eventTypes = AccessibilityEvent.TYPE_WINDOWS_CHANGED or
                AccessibilityEvent.TYPE_WINDOW_STATE_CHANGED or
                AccessibilityEvent.TYPE_WINDOW_CONTENT_CHANGED
        info.feedbackType = AccessibilityServiceInfo.FEEDBACK_ALL_MASK
        serviceInfo = info
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun setUpBottomSheet() {
        val mContainerParams = WindowManager.LayoutParams()
        mContainerParams.type = WindowManager.LayoutParams.TYPE_ACCESSIBILITY_OVERLAY
        mContainerParams.format = PixelFormat.TRANSLUCENT
        mContainerParams.flags =
            mContainerParams.flags or (WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE
                    or WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS
                    or WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
        mContainerParams.width = WindowManager.LayoutParams.MATCH_PARENT
        mContainerParams.height = WindowManager.LayoutParams.MATCH_PARENT
        binding = DialogSelectedTextsBinding.inflate(LayoutInflater.from(this))
        binding!!.markableView.highlightedColor =
            ColorUtils.setAlphaComponent(binding!!.markableView.highlightedColor, 255 / 4)
        binding!!.markableView.setBackgroundColor(0)
        binding!!.markableView.setOnTouchListener(this)
        mGestureDetector = GestureDetector(this, this)
        try {
            mWindowManager!!.addView(binding!!.markableView, mContainerParams)
        } catch (e: Exception) {
            val permissionGiven = AccessibilityServiceUtils.isAccessibilityServiceEnabled(
                this, ScreenTranslateService::class.java
            )
            Utils.toast(this, windows.toString())
            Log.e(TAG, e.toString())
            Log.i(TAG, permissionGiven.toString())
        }

        BottomSheetBehavior.from(binding!!.sheetContent).state = BottomSheetBehavior.STATE_COLLAPSED
        binding!!.imgClose.setOnClickListener { stopTranslateService() }
        binding!!.imgCopy.setOnClickListener {
            if (mSelectedTexts!!.isEmpty()) {
                Utils.showSnackbar(binding!!.markableView,"No texts were selected . Click on texts you want copy")
                return@setOnClickListener
            }
            val sb = StringBuilder()
            mSelectedTexts!!.forEach { sb.append(it + "\n") }
            Utils.copyToClipboard(this, sb)
            Utils.showSnackbar(binding!!.markableView, "Copied to clipboard")
        }
        binding!!.imgTranslate.setOnClickListener {
            if (mSelectedTexts!!.isEmpty()) {
                Utils.showSnackbar(binding!!.markableView,"No texts were selected . Click on texts you want translate")
                return@setOnClickListener
            }else {
                val intent = newIntent(this, mSelectedTexts!!)
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                startActivity(intent)
                stopTranslateService()
            }
        }
        binding!!.imgSelectAll.setOnClickListener {
            val list: List<AccessibilityNodeInfo> = ArrayList()
            AccessibilityServiceUtils.getAllNodesWithText(this, rootInActiveWindow)
            for (info in list) {
                addNode(info)
            }
        }
    }

    private fun setUpRecyclerView() {
        val rv = binding!!.recyclerview
        rv.layoutManager = LinearLayoutManager(this)
        mSelectedTextAdapter = SelectedTextsAdapter(mSelectedTexts)
        rv.adapter = mSelectedTextAdapter
        mSelectedTextAdapter!!.setOnCloseClickListener { pos: Int -> removeNode(mSelectedNodes!![pos]) }
    }

    private fun setUpContainer() {


    }

    private fun addNode(accessibilityNodeInfo: AccessibilityNodeInfo) {
        if (mSelectedNodes!!.contains(accessibilityNodeInfo)) return
        mSelectedNodes!!.add(accessibilityNodeInfo)
        val rect = Rect()
        accessibilityNodeInfo.getBoundsInScreen(rect)
        binding!!.markableView.addRect(rect)
        mSelectedTexts!!.add(accessibilityNodeInfo.text.toString())
        mSelectedTextAdapter!!.notifyItemInserted(mSelectedTexts!!.size - 1)
        refreshSelectedCount()
    }

    private fun removeNode(node: AccessibilityNodeInfo) {
        val index = mSelectedNodes!!.indexOf(node)
        if (index != -1) {
            val rect = Rect()
            node.getBoundsInScreen(rect)
            binding!!.markableView.removeRect(rect)
            mSelectedNodes!!.removeAt(index)
            mSelectedTexts!!.removeAt(index)
            mSelectedTextAdapter!!.notifyItemRemoved(index)
            mSelectedTextAdapter!!.notifyItemRangeChanged(index, mSelectedTextAdapter!!.itemCount)
            refreshSelectedCount()
        }
    }

    @SuppressLint("SetTextI18n")
    private fun refreshSelectedCount() {
        val prefix = if (mSelectedTexts!!.size == 0) "No text" else mSelectedTexts!!.size.toString()
        binding!!.selectedTextCount.text = "$prefix  selected"
    }

    private fun getTexts(node: AccessibilityNodeInfo?, x: Int, y: Int): MutableList<AccessibilityNodeInfo> {
        val list : MutableList<AccessibilityNodeInfo> = mutableListOf()
        if (node == null) return list
        if (node.text != null) {
            val rect = Rect()
            node.getBoundsInScreen(rect)
            if (rect.contains(x, y)) {
                list.add(node)
            }
        }
        for (i in 0 until node.childCount) {
            val child = node.getChild(i)
            list.addAll(getTexts(child, x, y))
        }
        return list
    }

    private fun stopTranslateService() {
        mWindowManager!!.removeView(binding!!.markableView)
        serviceInfo = AccessibilityServiceInfo()
        mSelectedWindowId = 0
        mWindowManager = null
        binding = null
        mGestureDetector = null
        mSelectedTexts = null
        mSelectedNodes = null
    }

    override fun onAccessibilityEvent(event: AccessibilityEvent) {
            if (mSelectedWindowId != 0) {
                // Window changed so stop the service
                if (mSelectedWindowId != event.windowId) {
                    stopTranslateService()
                }
            }
    }

    override fun onDestroy() {
        super.onDestroy()
        (getSystemService(NOTIFICATION_SERVICE) as NotificationManager).cancel(NOTIFICATION_ID)
        Log.v(TAG, "Service Destroyed")
        Utils.toast(this, "destroyed")
    }

    override fun onInterrupt() {
        Log.v(TAG, "onInterrupt")
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouch(v: View, e: MotionEvent): Boolean {
        return mGestureDetector!!.onTouchEvent(e)
    }

    override fun onLongPress(e: MotionEvent) {}
    override fun onScroll(e1: MotionEvent, e2: MotionEvent, distanceX: Float, distanceY: Float): Boolean = false
    override fun onFling(e1: MotionEvent, e2: MotionEvent, velocityX: Float, velocityY: Float): Boolean = false
    override fun onShowPress(e: MotionEvent) {}
    override fun onDown(e: MotionEvent): Boolean = false

    companion object {
        private const val TAG = "ScreenTranslateService"
        const val NOTIFICATION_ID = 100101
        const val ACTION_HIDE_NOTIFICATION = "hide_notification"
        const val ACTION_SHOW_NOTIFICATION = "show_notification"
        const val ACTION_START_TRANSLATE_MODE = "start_translate_mode"
    }
}